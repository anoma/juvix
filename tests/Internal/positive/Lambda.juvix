module Lambda;

open import Stdlib.Prelude public;

id' : {A : Type} → A → A;
id' := λ { a := a };

uncurry' : {A : Type} → {B : Type} → {C : Type} → (A → B → C) → A × B → C;
uncurry' := λ {f (a, b) := f a b};

fst' : {A : Type} → {B : Type} → A × B → A;
fst' {_} := λ {(a, _) := a};

first' : {A : Type} → {B : Type} → {A' : Type} → (A → A') → A × B → A' × B;
first' := λ {f (a, b) := f a, b};

foldr' : {A : Type} → {B : Type} → (A → B → B) → B → List A → B;
foldr' := λ {_ z nil := z;
             f z (h ∷ hs) := f h (foldr' f z hs)};

main : IO;
main := printNatLn (id' zero)
        >> printNatLn (uncurry' (+) (one, one))
        >> printNatLn (fst' (zero, one))
        >> printNatLn (fst (first' ((+) one) (one, zero)))
        >> printNatLn (foldr' (+) zero (one ∷ two ∷ three ∷ nil));

end;
