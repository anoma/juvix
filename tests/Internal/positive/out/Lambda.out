zero
suc (suc zero)
zero
suc (suc zero)
suc (suc (suc (suc (suc (suc zero)))))
